### PR TITLE
Fix Bug 1479817, anchor link headings are obscured by toc

### PR DESCRIPTION
**Pull request by schalkneethling**
This resolves the problem described in bug 1479817 on Bugzilla:
https://bugzilla.mozilla.org/show_bug.cgi?id=1479817

I also extracted the `scrollToHeading` function into utils and removed the jQuery dependency it had. Also, added `scroll-behaviour: smooth;` to the `html` element so, now in supported browsers, the scroll behaviour will be silky smooth. (Try Fx)

@jwhitlock @hobinjk r?

### CODE REVIEW

File: kuma/static/js/components/local-anchor.js [hunk 1]
```diff
@@ -5,12 +5,21 @@
...
 
+    var mainDocument = document.getElementById('document-main');
     // collect all headings with an `id` attribute
     var headings = Array.from(document.querySelectorAll('h2[id]')).concat(
         Array.from(document.querySelectorAll('h3[id]'))
     );
     var pageUrl = document.location.href;
 
+    // if the url contains a hash, strip it away
+    if (window.location.hash) {
```
[the remark this answers is not in the document]
Comment (schalkneethling) [reply kuma/static/js/components/local-anchor.js]: > I'm not familiar with the APIs. What is the difference between document.location and window.location? Why do you switch between the two?

No difference, let me stick with document for consistency.

**Comment by schalkneethling** [comment]
Thanks for the review @jwhitlock - updated

File: kuma/static/js/components/local-anchor.js [hunk 2]
```diff
@@ -5,11 +5,17 @@
...
     // collect all headings with an `id` attribute
     var headings = Array.from(document.querySelectorAll('h2[id]')).concat(
         Array.from(document.querySelectorAll('h3[id]'))
     );
-    var pageUrl = document.location.href;
+    var pageUrl = new URL(document.location.href);
+
+    // if the url contains a hash, strip it away
+    if (pageUrl.hash) {
+        pageUrl = pageUrl.href.substr(0, window.location.href.indexOf('#'));
```
[the remark this answers is not in the document]
Comment (schalkneethling) [reply kuma/static/js/components/local-anchor.js]: Yup, this just needed a rebase.

**Comment by schalkneethling** [comment]
Ok, this was badly in need of a rebase. Think it was simmering here for a while ;)

Regarding the new bug > Actual result: "Topic" heading is obscured below the TOC.

This is another case we need to handle for in content links that links to a hash. On it.

**Comment by schalkneethling** [comment]
Additional issues addressed. Also found an additional problem(and fixed it) where the `svg` or the nested `path` would be the `target` of the click event.

Setting `pointer-events` to `none` prevents this from happening.

**Comment by schalkneethling** [comment]
Thanks for the review @jwhitlock and @hobinjk

File: kuma/static/js/components/local-anchor.js [hunk 3]
```diff
@@ -7,17 +7,25 @@
     'use strict';
 
     function prependLocalAnchors() {
-        // collect all headings in wiki articles with an `id` attribute
-        var headings = Array.from(document.querySelectorAll('#wikiArticle h2[id]')).concat(
-            Array.from(document.querySelectorAll('#wikiArticle h3[id]'))
+        var mainDocument = document.getElementById('document-main');
+        // collect all headings with an `id` attribute
+        var headings = Array.from(document.querySelectorAll('h2[id]')).concat(
```
[the remark this answers is not in the document]
Comment (schalkneethling) [reply kuma/static/js/components/local-anchor.js]: I can definitely scope is to the main document instead. Good catch.

**Comment by schalkneethling** [comment]
> This seems to work, but it removes the floating table of contents. Can we have both?

@jwhitlock When I have devtools open, the stickiness of the toc does go away but, I have not seen it not be sticky when devtools are closed. Are you seeing it not be sticky event with devtools closed?

**Comment by jwhitlock** [comment]
I wasn't aware that the TOC went away with dev tools open, thanks. Now the TOC is back.

Test #1 is still broken. If I go directly to http://localhost:8000/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach#Using_thisArg, then heading is hidden behind the TOC.

Test #2 works. If I go to http://localhost:8000/en-US/docs/MDN/Contribute/Howto/Tag, then click "Tagging problems you can fix", then the "Topic" link, I see the heading.  However, if I open a new window/tab with links with anchor tags, the heading is obscured:

* http://localhost:8000/en-US/docs/MDN/Contribute/Howto/Tag#Tagging_problems_you_can_fix
* http://localhost:8000/en-US/docs/MDN/Contribute/Howto/Tag#Topic

**Comment by schalkneethling** [comment]
> Test #1 is still broken. If I go directly to http://localhost:8000/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach#Using_thisArg, then heading is hidden behind the TOC.

Hmmm, I cannot reproduce this one, let me take another look.

> However, if I open a new window/tab with links with anchor tags, the heading is obscured:

Might be another edge case, let me look at this. Thanks @jwhitlock

**Comment by schalkneethling** [comment]
@jwhitlock Thanks for the review. Just to confirm, the issue below is the only remaining issue?

> I still see issues going directly to the anchor link in FF Dev Edition 63.0b11, if > I open in a new tab or window. The content immediately jumps to the heading, >hidden behind the TOC.

**Comment by jwhitlock** [comment]
I think so. [Bug 1479817](https://bugzilla.mozilla.org/show_bug.cgi?id=1479817) is specifically about what happens when you share a link with an anchor, such as over email or stack overflow, so I'd like for it to show the header more reliably.

I also don't want other stuff to break:
* TOC should be shown (minus issues around having dev tools open)
* Clicking the link icon should load the anchor into the location bar, not hide the header
* Clicking an anchor link in the content should go to that anchor, show the header
* Going to a URL without a hash component should still work

**Comment by schalkneethling** [comment]
Hey @jwhitlock 

I have tested this last scenario multiple times in Fx and Chrome and I cannot seem to reproduce the problem. For me it works as expected. See the videos below:

Firefox: https://cl.ly/9e4df462f7d5
Chrome: https://cl.ly/6e542d9b404c (It does a bit of a peekaboo but settles down correctly :))

And then just clicking around using the TOC in Fx
https://cl.ly/8d11bdad66af

**Comment by jwhitlock** [comment]
I still get browser differences, and differences between short mode (Table of Contents is non-sticky) and tall mode (Table of Contents is sticky).Here's my results:

* Firefox, low DPI, ~980 px, this PR: https://cl.ly/1a597ce498ee
  - [#ThisArg](http://localhost:8000/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach#Using_thisArg): Heading is top of page, no table of contents
  - [#Tagging_problems_you_can_fix](http://localhost:8000/en-US/docs/MDN/Contribute/Howto/Tag#Tagging_problems_you_can_fix): Heading is top of page, no table of contents
  - [#Topic](http://localhost:8000/en-US/docs/MDN/Contribute/Howto/Tag#Topic): Heading is top of page, no table of contents
  - [Howto/Tag](http://localhost:8000/en-US/docs/MDN/Contribute/Howto/Tag): Link in TOC goes to top of the heading, no table of contents. In-content anchor link pushes down heading as if there was a table of contents.
* Firefox, high DPI, ~980px, this PR:
  - [#ThisArg](http://localhost:8000/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach#Using_thisArg):
  - [#Tagging_problems_you_can_fix](http://localhost:8000/en-US/docs/MDN/Contribute/Howto/Tag#Tagging_problems_you_can_fix): 
  - [#Topic](http://localhost:8000/en-US/docs/MDN/Contribute/Howto/Tag#Topic):
  - [Howto/Tag](http://localhost:8000/en-US/docs/MDN/Contribute/Howto/Tag): 
* Chrome, low DPI, ~980 px, this PR: https://cl.ly/7e67c33010c8
  - [#ThisArg](http://localhost:8000/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach#Using_thisArg): Heading leaves room for table of contents, no table of contents
  - [#Tagging_problems_you_can_fix](http://localhost:8000/en-US/docs/MDN/Contribute/Howto/Tag#Tagging_problems_you_can_fix): Heading leaves room for table of contents, no table of contents
  - [#Topic](http://localhost:8000/en-US/docs/MDN/Contribute/Howto/Tag#Topic): Heading leaves room for table of contents, no table of contents
  - [Howto/Tag](http://localhost:8000/en-US/docs/MDN/Contribute/Howto/Tag): Link in TOC goes to top of the heading, no table of contents. In-content anchor link pushes down heading as if there was a table of contents.
* Chrome, high DPI, ~980px, this PR:
  - [#ThisArg](http://localhost:8000/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach#Using_thisArg):
  - [#Tagging_problems_you_can_fix](http://localhost:8000/en-US/docs/MDN/Contribute/Howto/Tag#Tagging_problems_you_can_fix): 
  - [#Topic](http://localhost:8000/en-US/docs/MDN/Contribute/Howto/Tag#Topic):
  - [Howto/Tag](http://localhost:8000/en-US/docs/MDN/Contribute/Howto/Tag):
* Firefox, low DPI, ~980 px, production:
  - [#ThisArg](http://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach#Using_thisArg):
  - [#Tagging_problems_you_can_fix](http://developer.mozilla.org/en-US/docs/MDN/Contribute/Howto/Tag#Tagging_problems_you_can_fix):
  - [#Topic](http://developer.mozilla.org/en-US/docs/MDN/Contribute/Howto/Tag#Topic):
  - [Howto/Tag](http://developer.mozilla.org/en-US/docs/MDN/Contribute/Howto/Tag):
* Firefox, high DPI, ~980px, production:
  - [#ThisArg](http://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach#Using_thisArg):
  - [#Tagging_problems_you_can_fix](http://developer.mozilla.org/en-US/docs/MDN/Contribute/Howto/Tag#Tagging_problems_you_can_fix): 
  - [#Topic](http://developer.mozilla.org/en-US/docs/MDN/Contribute/Howto/Tag#Topic):
  - [Howto/Tag](http://developer.mozilla.org/en-US/docs/MDN/Contribute/Howto/Tag):
* Chrome, low DPI, ~980 px, production:
  - [#ThisArg](http://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach#Using_thisArg):
  - [#Tagging_problems_you_can_fix](http://developer.mozilla.org/en-US/docs/MDN/Contribute/Howto/Tag#Tagging_problems_you_can_fix): 
  - [#Topic](http://developer.mozilla.org/en-US/docs/MDN/Contribute/Howto/Tag#Topic):
  - [Howto/Tag](http://developer.mozilla.org/en-US/docs/MDN/Contribute/Howto/Tag):
* Chrome, high DPI, ~980px, production:
  - [#ThisArg](http://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach#Using_thisArg):
  - [#Tagging_problems_you_can_fix](http://developer.mozilla.org/en-US/docs/MDN/Contribute/Howto/Tag#Tagging_problems_you_can_fix): 
  - [#Topic](http://developer.mozilla.org/en-US/docs/MDN/Contribute/Howto/Tag#Topic):
  - [Howto/Tag](http://developer.mozilla.org/en-US/docs/MDN/Contribute/Howto/Tag):

**Comment by jwhitlock** [comment]
Oops, did not mean to post that last comment, or close the PR.

**Comment by jwhitlock** [comment]
*Here's the comment I meant to post, once I discovered that the height of the window mattered*

I still get browser differences, and differences between short mode (Table of Contents is non-sticky) and tall mode (Table of Contents is sticky). The browser differences may be due to evaluation order of non-combined JS files, or if ``analytics.js`` is blocked. Here are my results:

* Firefox - https://cl.ly/31154afd856e
  - ✅ Short mode, [#Hash link](http://localhost:8000/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach#Using_thisArg): Jumps to correctly positioned heading 
  - ✅ Short mode, [Link without hash](http://localhost:8000/en-US/docs/MDN/Contribute/Howto/Tag), then click Table of Contents: Smooth scroll to correctly positioned heading
  - ❌ Short mode, click on-page anchor link: Smooth scroll to header, but header includes space for missing table of contents. 
  - ❌ ❓ Tall mode, [#Hash link](http://localhost:8000/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach#Using_thisArg): Jumps to header hidden behind sticky table of contents. This is different from @schalkneethling 's results in https://cl.ly/9e4df462f7d5, where it jumps to the correctly-positioned header
  - ✅ Tall mode, [Link without hash](http://localhost:8000/en-US/docs/MDN/Contribute/Howto/Tag), then click Table of Contents: Smooth scroll to correctly positioned header. Same as @schalkneethling's https://cl.ly/8d11bdad66af.
  - ✅ Tall mode, click on-page anchor link: Smooth scroll to correctly positioned header  
* Chrome - https://cl.ly/7a262e47b86a. Same as @schalkneethling's https://cl.ly/8d11bdad66af.
  - ✅ / ❌ Short mode, [#Hash link](http://localhost:8000/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach#Using_thisArg): First link, smooth scroll from top of page to the correctly positioned header. On second and third links, smooth scroll, but header includes space for missing table of contents .
  - ✅ Short mode, [Link without hash](http://localhost:8000/en-US/docs/MDN/Contribute/Howto/Tag), then click Table of Contents: Smooth scroll to correctly positioned header.
  - ❌ Short mode, click on-page anchor link: Smooth scroll to header that includes space for missing table of contents. 
  - ✅ Tall mode, [#Hash link](http://localhost:8000/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach#Using_thisArg): Smooth scroll from top of page to the correctly positioned header. Same as @schalkneethling's results - https://cl.ly/6e542d9b404c
  - ✅ Tall mode, [Link without hash](http://localhost:8000/en-US/docs/MDN/Contribute/Howto/Tag), then click Table of Contents: Smooth scroll from top of page to the correctly positioned header  
  - ✅ Tall mode, click on-page anchor link: Smooth scroll to the correctly positioned header  

I think the issue with Tall mode (with sticky ToC) and direct hash links is going to be hard to reproduce, and is possibly a JS ordering issue. We should retest when it is on staging or production, but not block.

I think more work could be done to only include the Table of Content padding when the browser is tall enough for a sticky ToC, which reproduces for me across browsers. @schalkneethling, if you think that is in scope for this change, please try to reproduce and adjust the code.

**Comment by schalkneethling** [comment]
Thanks for the detailed feedback @jwhitlock . I am on it.

**Comment by schalkneethling** [comment]
@jwhitlock Whoah ok, so the calculation results by Chrome and Firefox are pretty different. Seeing that Chrome is the one that gets it right, I assume the problem lies with Fx. I am going to hunt down for a known bug, or other properties where the difference is not so extreme:

Firefox:

headingTop 1122.4833374023438 (Difference of ~95)
tocHeight 48
scrollY 1074.4833374023438 (Difference of ~96)

Chrome:

headingTop 1218
tocHeight 48
scrollY 1170

**Comment by schalkneethling** [comment]
w00t! Thanks for the thorough review @jwhitlock

**Pull request merged**